### DISCUSSION TITLE
Orientation tag changes during exif copy? Or not?

### DISCUSSION THREAD

**Discussion by lphilpot** [Exiv2/exiv2#2168] [Q&A] [answered]
_exiv2 0.27.5 on Windows 11_

I'm new to exiv2 and not intimately familiar with exif (beyond reading the values) so I'm not sure exactly what to ask nor what else I'm doing wrong. I'm trying to copy tags from one file to another (a CR3 to a jpg in this case) because RawTherapee doesn't yet fully handle CR3 metadata. I'm doing this, which I read at https://dev.exiv2.org/boards/3/topics/2329 :

`exiv2 -PkV filename.CR3 | exiv2 -m- filename.jpg`

That does copy the tags, but it also changes (per XnView at least) the orientation value from 1 to 8. Indeed, the JPG gets rotated 90 degrees CCW. FastStone shows it rotated and opening it in GIMP I'm prompted whether to rotate or not. However, exiv2 apparently doesn't see any difference between the two files:

```
C:\data\images\Our-photos\_scratch\tmp>exiv2 --grep Orientation IMG_0554.CR3
Exif.Image.Orientation                       Short       1  left, bottom

C:\data\images\Our-photos\_scratch\tmp>exiv2 --grep Orientation IMG_0554.jpg
Exif.Image.Orientation                       Short       1  left, bottom
```

I've not yet fully gotten my head around the exiv syntax so experimentation hasn't gotten me too far. Can someone explain what's happening, how I can fix this and / or if there's a better way to copy tags? I don't want to modify them, I just want populate the tags RawTherapee is skipping.

Thanks!

**Comment by clanmills** [comment]
I think you've got your head around the command-line program.  It's a bit of swiss army knife, however you've got the basic idea.   Piping the metadata is probably more "advanced" than you need yet.   Additionally, I think a bug was reported and fixed concerning the piping code.  So, let's focus on using "simple" commands.  The most basic commands are:

1 Reading metadata value

```
553 rmills@rmillsm1:~/Pictures/Photos $ exiv2 -pv --grep Orientation ~/Stonehenge.jpg 
0x0112 Image        Orientation                 Short       1  1
554 rmills@rmillsm1:~/Pictures/Photos $ 
```

This says "The tag for Orientation is 0x0112 in the Tiff Specification, and is an array with one 16-bit integer which as a value of 1".

It's more common to print the "translated values" (translated = interpreted for human use)

```
554 rmills@rmillsm1:~/Pictures/Photos $ exiv2 -pt --grep Orientation ~/Stonehenge.jpg 
Exif.Image.Orientation                       Short       1  top, left
555 rmills@rmillsm1:~/Pictures/Photos $ 
```
The output is similar, however instead of showing the value of "1", it is "translated" to the string "top, left".


2.  Setting a value

```
559 rmills@rmillsm1:~/Pictures/Photos $ cp ~/Stonehenge.jpg .
560 rmills@rmillsm1:~/Pictures/Photos $ exiv2 -M'set Exif.Image.Orientation 4' Stonehenge.jpg 
561 rmills@rmillsm1:~/Pictures/Photos $ exiv2 -pv --grep Orientation Stonehenge.jpg 
0x0112 Image        Orientation                 Short       1  4
562 rmills@rmillsm1:~/Pictures/Photos $ exiv2 -pt --grep Orientation Stonehenge.jpg 
Exif.Image.Orientation                       Short       1  bottom, left
563 rmills@rmillsm1:~/Pictures/Photos $ 
```

@postscript-dev has done very good work on the documentation on the 'main' branch and I highly recommend that you read that.

The `main` branch is a major refactoring of the code into C++17 and this effort has been in progress since 2018.  While that is being done, I have been making "dot" releases on branch '0.27-maintenance'.  Functionally, both 'main' and '0.27-maintenance' are similar.  However I don't remember if the fix concerning piping was back-ported to 0.27-maintenance.

Please be aware that applications which display images on your computer might not understand the Orientation tag.

I am happy to meet with you on Zoom to chat about the exiv2 command-line utility.

**Comment by lphilpot** [comment]
_(Apologies in advance for the long post)_

Thank you very much for the prompt, detailed reply and for the generous Zoom offer! At this point, I'm mainly trying to devise a method that will reliably just copy tags from one file to another. As mentioned, RawTherapee (which I use from time to time) doesn't yet handle CR3 metadata, so in order to have EXIF in my processed images I need to manually copy it from the raw file. Also, I've had occasional instances where GIMP doesn't write EXIF to exported images. On my previous computer (Xubuntu 20.04) I wrote a script that used ExifTool to batch process images but it was suggested exiv2 might be a better option, hence my research here. However I'm confused what's actually going on here ...no surprise there!  :-)

Data points, FWIW...

Prior to copying tags:
```
C:\data\images\Our-photos\_scratch\tmp>exiv2 --grep Orientation IMG_0554.CR3
Exif.Image.Orientation                       Short       1  left, bottom
C:\data\images\Our-photos\_scratch\tmp>exiv2 --grep Orientation IMG_0554.jpg
```
i.e., no EXIF in the JPG.

After copying tags with the above exiv2 command syntax:
```
C:\data\images\Our-photos\_scratch\tmp>exiv2 --grep Orientation IMG_0554.CR3
Exif.Image.Orientation                       Short       1  left, bottom
C:\data\images\Our-photos\_scratch\tmp>exiv2 --grep Orientation IMG_0554.jpg
Exif.Image.Orientation                       Short       1  left, bottom
```
But in _almost_ every app I view the JPG, it's rotated 270 degrees. And you're right, there can certainly be differences / errors in how apps interpret tags. For example:

**XnView**
- CR3 portrait, JPG 270 landscape
- EXIF panel (Camera | Orientation) shows "top-left (1)" for CR3 but "left-bottom(8)" for JPG
- ExifTool panel (EXIF | Orientation) shows "Rotate 270 clockwise" for both files, which agrees with CLI output from ExifTool.

_Indeed, the shot was taken in portrait orientation but I'm 99% sure it was with the camera rotated 270 counter/anti-clockwise, not clockwise. Maybe whatever ExifTool is reading as "Rotate 270 clockwise" is being interpreted by others to incorrectly rotate it yet again?_

[ Update - Duhh... same thing: left 90 vs right 270. It's a slow Sunday afternoon, obviously! :-) ]

**FastStone Image Viewer**
- CR3 portrait, JPG 270 landscape; metadata panel doesn't show orientation tags

**GIMP**
- Doesn't open the CR3, but says the JPG contains orientation info and I'm prompted to rotate or not.

**RawTherapee**
- CR3 portrait, 270 JPG landscape

**darktable**
- CR3 portrait, JPG **_portrait_**
- This is the odd-man-out, but unfortunately the image properties module doesn't include Orientation.

So what's happening? I can't tell. I can attach the files pre- or post-copy, if desired.

It may be an edge case, but I'd like to handle edge cases properly. At this point I'm more interested in getting the copy process correct and will deal with individual image issues if / when they arise. 

Can you point me to a proper, reliable syntax? Or are the steps you outlined above as concise as it currently gets?

Thanks!

**Reply by clanmills** [reply]
I don't use any of those applications, so I can't say anything about them.

Until recently, I never took "portrait" photos.  The reason is because they disturb the layout of photos on my web-site: clanmills.com.  All landscape looks like this:  https://clanmills.com/2022/Trips/Rome/Cupola/  Recently, the web-site has been re-engineered to use justifiedGallery which handles a mix of portrait and landscape with grace like this:  https://clanmills.com/2022/Trips/Rome/VaticanMuseum/

What's going on?  I've explained what's going on in Exiv2.  This week I will investigate the status of the back-port to 0.27-maintenance of the "piping fix".  Did you build exiv2 from source, or have you downloaded the build from exiv2.org?

Team Exiv2 has a friendly relationship with GIMP, darktable and RT.  I've heard of XnView and not of FastStone.  I'm a retired Adobe Engineer (PostScript, ExtendScript, Acrobat) and believe Adobe products handle Orientation correctly.  I personally use ACDSee Photo Studio 8 for Mac to manage my 80,000 photos and it handles Orientation.  In my photos I took in Rome last week, I've used Orientation 1, 6 and 8 and ACDSee (and Apple's Preview.app) display them correctly.

Perhaps, we could both focus this discussion on _**darktable**_.  You've said _image properties module doesn't include Orientation_.  Does this matter?  If _**darktable**_ understands Orientation, why do you need to change it?  I am happy to talk/screen-share on Zoom and to invite somebody from _**darktable**_ to join us.

**Comment by clanmills** [comment]
I've just thought of something totally different.  My camera (a Nikon D5300) has a setting "Auto Image Rotation" ON/OFF.  It's set to "ON". What does it do?  The User Manual doesn't say.  _I suspect it concerns the preview display on the back of the camera._

It's possible to have "portrait" photos with Orientation = 1 by writing the image with more rows of shorter length.  I believe my camera _**always**_ writes the image with more pixels/rows and uses the information from the rotation sensor to determine Exif.Image.Orientation 

However, there it's possible for the camera to use the information from the rotation sensor to determine how the image is to be written.

And there's a third possibility.  Your camera doesn't have a rotation sensor.

**Reply by clanmills** [reply]
I have investigated the D5300 "Auto Image Rotation" ON/OFF feature.  It is a camera UI feature which is not discussed in the User Manual.  However it is documented in "Reference Manual" p179.
https://download.nikonimglib.com/archive3/2Pnka00a3Hjt04DYIAO00pxxNF72/D5300VRRM_(En)04.pdf

**Comment by postscript-dev** [comment]
@lphilpot:

The `main` branch manpage is available [here](https://github.com/Exiv2/exiv2/blob/main/exiv2.md) and contains more information and examples. At the moment, there is very little that only applies to the `main` branch (see the file's history).

@clanmills is correct that there are a small number of bugs in the `exiv2` application, you can find some listed in https://github.com/Exiv2/exiv2/issues/1934 . I have already fixed some on `main` and when I find the time, I plan to submit a PR for the others. Bug 4 will be of interest to you as using [--extract tgt3](https://github.com/Exiv2/exiv2/blob/main/exiv2.md#extract_tgt3) and [--insert tgt3](https://github.com/Exiv2/exiv2/blob/main/exiv2.md#insert_tgt2) is the typical way in which you would copy all the metadata in a file. As you found, the workaround is to use the [--Modify cmd](https://github.com/Exiv2/exiv2/blob/main/exiv2.md#Modify_cmd) command but this has limitations. For example, this does not copy the thumbnail image or ICC profile to the new image. Bug 2 explains that you cannot stream the thumbnail but a workaround is to extract the thumbnail to a file and then insert it. Bug 2 has been fixed in the `main` branch.

I would make one small change to your command line though as when printing non built-in `Xmp` tagnames (see [key syntax](https://github.com/Exiv2/exiv2/blob/main/exiv2.md#exiv2_key_syntax)). the type needs to be specified. This is because the default for `Xmp` is `XmpText` and any [multiple entry types](https://github.com/Exiv2/exiv2/blob/main/exiv2.md#xmp-2) such as `XmpBag` (e.g., `"item 1", "item 2", "item 3"`) will be flattened into a single entry `XmpText` item (e.g., `"item 1, item 2, item 3"`). The solution is to add the type when printing. For example,

```bash
$ exiv2 -PkyV filename.CR3 | exiv2 -m- filename.jpg
```
This bug has been fixed in the `main` branch.

As for the individual applications, @clanmills has come up with some good suggestions but it is difficult to investigate more without a test image.

**Reply by clanmills** [reply]
Thank You @postscript-dev Peter.  As always, what you have said is very helpful.

We need to narrow the "depth of field" here.  What's happening in the metadata is straight forward.  The tag Exif.Image.Orientation is has one of 8 values (1-8) as specified in the Tiff Spec.

The exiv2 command-line feature "metadata piping" isn't as strong as it could be.  I've asked  @lphilpot Len not to use that.

The area of ambiguity is the behaviour of desktop applications.  We can simplify that if everybody uses darktable.

Because Len's email arrived about 23:30 BST on Sunday Evening, I suspect @lphilpot Len is in Australia or NZ.  One of the darktable engineers (Matt McG) is in Australia and is very helpful.  If I talk to Len on Zoom, I'll invite Matt to join us.

**Comment by clanmills** [comment]
Another interesting discovery.  "Auto Image Rotation" ON/OFF on the D5300 causes the image to be written "side-ways" with Exif.Image.Orientation = 1.  ACDSee _correctly_ displays the image "on its side".  I fixed it with the command:

```bash
$ exiv2 -M'set Exif.Image.Orientation 8' dsc_0003.jpg
```

I was little disappointed that ACDSee doesn't monitor the FileSystem and update the UI when the file was modified.  However leaving and re-entering the folder worked perfectly.

**Comment by lphilpot** [comment]
@clanmills 

> I don't use any of those applications, so I can't say anything about them.

I mentioned them only because they (and exiv2, exiftool) provide me visibility into EXIF data.

> Did you build exiv2 from source, or have you downloaded the build from exiv2.org?

I downloaded a Windows binary (exiv2-0.27.5-2019msvc64.zip) from exiv2.org, I believe. Until two months ago I ran Linux at home (for ~6 years) but my new laptop has Windows 11 on it so I'm getting re-familiar with Windows again. I've not yet set up a dev/build environment on it.

> Does this matter? If darktable understands Orientation, why do you need to change it?

I don't need to change it. As indicated, I mentioned darktable simply as one of several auxiliary references / test cases. The apparent change in target-image orientation when viewed in the other apps is what triggered my questions here.

My intent is to simply populate a few tags if missing in an image -- Typically due to RawTherapee not yet fully supporting CR3 metadata or whatever occasionally happens in GIMP.

> However, there it's possible for the camera to use the information from the rotation sensor to determine how the image is to be written.
> And there's a third possibility. Your camera doesn't have a rotation sensor.

I have a Canon T8i (850D). I've not found it documented anywhere but it appears to detect orientation changes. I shot a series of four images at 90 degree intervals: portrait, rotated 90CW, inverted, rotated 270 CW. All display correctly, all are full size (not truncated), etc. Both exiv2 and exiftool show they've been rotated:

exiv2
```
Exif.Image.Orientation   Short       1  top, left
Exif.Image.Orientation   Short       1  left, bottom
Exif.Image.Orientation   Short       1  top, left
Exif.Image.Orientation   Short       1  right, top
```
exiftool
```
Orientation    : Horizontal (normal)
Orientation    : Rotate 270 CW
Orientation    : Horizontal (normal)
Orientation    : Rotate 90 CW
```

> I suspect @lphilpot Len is in Australia or NZ. One of the darktable engineers (Matt McG) is in Australia and is very helpful. If I talk to Len on Zoom, I'll invite Matt to join us.

I'm in the southern US (Louisiana), Central Time zone.

As I mentioned before I appreciate your Zoom offer but I think that's overkill for me at the moment. I'm not a developer (I wish had been, instead of a sysadmin!) and at present, not even up to speed on how tags are interpreted, etc. I'm just trying to work around a bug or two. I'm afraid I'd be way out of my depth in such a conversation and would be able to contribute nothing of value.

Plus, the last thing I want is to potentially waste anyone's time on a 'wild goose chase'. Other than what I mentioned about RT and GIMP, I have no reason to suspect misbehavior from any app (nor my camera) at this point.

Thanks again for all the info.

@postscript-dev 

Thank you for the info and recommendations. I'll take a look ASAP -- kinda tied up at the moment. I don't mind posting a test image if it comes to that, but I think first I need to do my homework and make sure I'm using exiv2 correctly.

**Comment by clanmills** [comment, marked as answer]
It's never a "wild goose chase" to help a user.  I consider User Support and Security to be my top priorities for Exiv2.  Peter (@postscript-dev) and I are always happy to help.  Being retired, I have more time. 

You have a very good camera which has a rotation sensor.  You've proven that with the metadata from ExifTool and Exiv2.  I think every Canon has a rotation sensor.  I have a little point-n-shoot Canon IXUS-185 ($99) which has a rotation sensor.

However, maybe playing with the apps and the exiv2 commands that you really understand will bring you to "enlightenment" and give you the confidence to work with other tags.

```
$ exiv2 -pv --grep Orientation path
$ exiv2 -M'set Exif.Image.Orientation 6' path
```

Setting up a developer environment on Windows?  WSL (Windows Subsystem for Linux) is Linux!  (not kidding, it is linux).   Cygwin and MinGW feel very much like Linux.  The "native" Windows Development environment is Visual Studio and M/S provide a free Community Edition.  If you are familiar with GCC and the Linux tools, use WSL.  Visual Studio is really good, however it will feel very different from Linux.

Louisiana?  Been in New Orleans and all 50 States.  Proud to be a US Citizen.  Just as proud to be Scottish and retired in England.  https://clanmills.com/2005/SouthernTrip/Homepage/

Good Luck.

**Comment by lphilpot** [comment]
> Setting up a developer environment on Windows? WSL (Windows Subsystem for Linux) is Linux! (not kidding, it is linux).

Yep, I've got WSL installed (Ubuntu) as well as Linux Mint running in a VirtualBox VM. But I was thinking about a native Windows build env. Years ago I bought Visual C++ 2003. I've also tried Code::Blocks / MinGW. 20+ years ago I played with C, VB, Perl, QuickBASIC, etc. at a hobbyist level. Never came to terms with C++ and OOP, but I _really_ found C interesting (still do). However I've not thought about it for many years (and "use it or lose it" is most definitely true). Since then I did some Bash scripting as part of my unix admin job, but on the Windows side I haven't really learned too much about PowerShell just yet (it's definitely from MS, no doubt).   :-)

So you saw NOLA only a few months before Katrina hit. I'm about 200 miles to the NE, in the center of the state. I'm probably not the most enthusiastic flag-waver for 'all things Louisiana', but I'll admit they've got that food thing nailed...

Thanks.

**Comment by clanmills** [comment]
Yes.  We watched Katrina from San Jose, CA.  Because of our visit, we understood the geography.   Natchez Trace Parkway is on our "bucket list" along with Selma, AL.  A visit to family in Houston, TX might be followed a trip near you.

When we visited the Jefferson Davis Library in Biloxi, I took a photo of the death mask.  A few years later I got an email from Rice University requesting the hi-res original.  Post Katrina, nobody knew where the death mask had been stored.  They wanted a photo for a book they were publishing:  https://jeffersondavis.rice.edu

I've never learnt PowerShell, however I know and use bash a lot.  Python is my favourite scripting language and the Exiv2 test suite is now 100% Python.

One of the reasons I worked on Exiv2 since retiring is to retain my C++ and cross-platform skills.  The other reason is the team of great people who work with me on Exiv2.  I'm finally on my way "out the door" on Exiv2.  The Team deal with the tough stuff.

**Reply by lphilpot** [reply]
If you define "learnt PowerShell" to include a working knowledge of the majority of its features, then neither have I, for sure. I wrote / adapted little one-liners and hacked up scripts snarfed off the web for work tasks, generally to try and do poorly on Windows what I could already do on Linux / Solaris with a shell script.  :-)  Now at home (retired) with no domain, there's not a lot of incentive to learn it.

Anyway, I don't need to turn github into a chat room, so thanks again and I'll holler if I run into issues.

**Reply by clanmills** [reply]
Thanks, I've enjoyed our chat.  Best of Luck.

**Comment by tester0077** [comment]
Hopefully you are aware of the excellent exif orientation sample files at:
https://github.com/recurser/exif-orientation-examples

as well as the detailed blog about the subject at:
https://www.daveperrett.com/articles/2012/07/28/exif-orientation-handling-is-a-ghetto/
